source.go

```diff
@@ -43,18 +56,215 @@ func (s *Source) Configure(ctx context.Context, cfg map[string]string) error {
+		return fmt.Errorf("failed parsing time string %v: %w", string(pos), err)
+	}
+	s.lastEditedTime = posParsed
 	return nil
 }
 
 func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
-	return sdk.Record{}, nil
+	err := s.populateIDs(ctx)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching page IDs: %w", err)
+	}
+	if len(s.fetchIDs) == 0 {
+		return sdk.Record{}, sdk.ErrBackoffRetry
+	}
+
+	return s.nextObject(ctx)
+}
+
+func (s *Source) nextObject(ctx context.Context) (sdk.Record, error) {
+	if len(s.fetchIDs) == 0 {
+		return sdk.Record{}, errors.New("no page IDs available")
+	}
+	pageID := s.fetchIDs[0]
+	s.fetchIDs = s.fetchIDs[1:]
+
+	// todo support databases
+	sdk.Logger(ctx).Debug().
+		Str("page_id", pageID).
+		Msg("fetching page")
+
+	block, err := s.client.Block.Get(ctx, notion.BlockID(pageID))
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching page %v: %w", pageID, err)
+	}
+
+	children, err := s.getChildren(ctx, block)
+	if err != nil {
+		return sdk.Record{}, fmt.Errorf("failed fetching blocks for %v: %w", pageID, err)
+	}
+
+	record, err := s.blockToRecord(block, children)
+	if err != nil {
+		return sdk.Record{}, err
+	}
+	s.lastEditedTime = *block.GetLastEditedTime()
+	return record, nil
+}
+
+// getChildren gets all the child and grand-child blocks of the input block
+func (s *Source) getChildren(ctx context.Context, block notion.Block) ([]notion.Block, error) {
+	var children []notion.Block
+	if !block.GetHasChildren() {
+		sdk.Logger(ctx).Debug().
+			Str("block_id", block.GetID().String()).
+			Msg("block has no children")
+		return children, nil
+	}
+
+	fetch := true
+	var cursor notion.Cursor
+	for fetch {
+		resp, err := s.client.Block.GetChildren(
+			ctx,
+			block.GetID(),
+			&notion.Pagination{
+				StartCursor: cursor,
+			},
+		)
+		if err != nil {
+			return nil, fmt.Errorf(
+				"failed getting children for block ID %v, cursor %v: %w",
+				block.GetID(),
+				cursor,
+				err,
+			)
+		}
+
+		// get grandchildren as well
+		for _, child := range resp.Results {
+			children = append(children, child)
+			grandChildren, err := s.getChildren(ctx, child)
+			if err != nil {
+				return nil, err
+			}
+			children = append(children, grandChildren...)
+		}
+
+		fetch = resp.HasMore
+		cursor = notion.Cursor(resp.NextCursor)
+	}
+	return children, nil
 }
 
-func (s *Source) Ack(ctx context.Context, position sdk.Position) error {
+func (s *Source) Ack(context.Context, sdk.Position) error {
 	return nil
 }
 
-func (s *Source) Teardown(ctx context.Context) error {
+func (s *Source) Teardown(context.Context) error {
 	return nil
 }
+
+func (s *Source) populateIDs(ctx context.Context) error {
+	if len(s.fetchIDs) > 0 {
+		return nil
+	}
+	// the first read attempt (when the connector starts)
+	if !s.firstFetch {
+		sdk.Logger(ctx).Debug().
+			Dur("poll_interval", s.config.pollInterval).
+			Msg("sleeping before checking for changes")
+		time.Sleep(s.config.pollInterval)
```

Comment:
For the first version this is.. well.. acceptable 😅 just know that since we are returning `ErrBackoffRetry` if there are no changes the SDK will already do some sleeping between calls to `Read`, so the poll interval is more like a "minimum delay between polls". Ideally we would have a background job that fetches these changes and feeds them to `Read` via a channel or something (the s3 connector does it that way).

Reply:
I guess the goal we want to achieve, is that the `Read` method doesn't block?

Reply:
> in the ideal world `Read` blocks until there's a record or the context gets cancelled

So returning an `ErrBackoffRetry` isn't actually recommended?

Reply:
I'll resolve this conversation to be able to merge the PR, but I'll be happy to any implement fixes/improvements here in a follow-up PR.

Reply:
Regarding `ErrBackoffRetry` I'd say it's just a utility that makes your life easier. It allows you to easily get started with polling, but you (the developer) are not in charge of the delay between the retries, so it's not ideal (well we have an [issue](https://github.com/ConduitIO/conduit/issues/184) for that). It's not _wrong_ to use it, but not the best either.

You also have to consider that there might be significant delays between calls to `Read` if the pipeline is "full" (I'm talking about backpressure in pipelines with slow destinations). In that case, we are losing precious time when we could be already downloading the next page and preparing it so that when `Read` is called we already have a record ready to return. So having a background job that downloads records and stores them in a cache makes the code a bit more complicated, but in turn it keeps the connector completely in charge of the polling interval and makes it more performant by downloading pages even while the SDK is busy returning a record back to Conduit.

Reply:
FTR: we'll implement it in a follow-up PR (tracking issue https://github.com/conduitio-labs/conduit-connector-notion/issues/12), so that Samir is unblocked in his work.